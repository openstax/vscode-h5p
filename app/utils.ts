import { FileInfo, Message } from './models';
import fsExtra from 'fs-extra';
import os from 'os';

// 2 GB = 2 * 1024 * 1024 * 1024 B
export const MAX_SIZE_IN_BYTES = 2147483648;

export function isFileInfoMessage(
  message: Message
): message is Message<FileInfo> {
  return message.type === 'FileInfo';
}

/**
 * This method will delete all temporary uploaded files from the request
 */
export async function clearTempFiles(
  req: Request & { files: any }
): Promise<void> {
  if (!req.files) {
    return;
  }

  await Promise.all(
    Object.keys(req.files).map((file) =>
      req.files[file].tempFilePath !== undefined &&
      req.files[file].tempFilePath !== ''
        ? fsExtra.remove(req.files[file].tempFilePath)
        : Promise.resolve()
    )
  );
}
