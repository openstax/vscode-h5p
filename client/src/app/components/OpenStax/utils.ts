export function* range(a: number, b?: number) {
  if (b === undefined) {
    for (let i = 0; i < a; i++) yield i;
  } else {
    for (let i = a; i < b; i++) yield i;
  }
}

export function collect<T>(iterable: Iterable<T>) {
  return [...iterable];
}

export function chunk<T>(arr: T[], chunkSize: number): Array<T[]> {
  return [...range(Math.ceil(arr.length / chunkSize))]
    .map((n) => n * chunkSize)
    .map((offset) => arr.slice(offset, offset + chunkSize));
}

export function debounce<A = unknown, R = void>(
  fn: (args: A) => R,
  ms: number
): (args: A) => Promise<R> {
  let timer: NodeJS.Timeout;

  return (args: A) => {
    clearTimeout(timer);

    return new Promise((resolve) => {
      timer = setTimeout(() => {
        resolve(fn(args));
      }, ms);
    });
  };
}
