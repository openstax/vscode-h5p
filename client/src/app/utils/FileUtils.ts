import { FileInfo, Message } from '../models/FileInfo';

export function isFileInfoMessage(
  message: Message,
): message is Message<FileInfo> {
  return message.type === 'FileInfo';
}
