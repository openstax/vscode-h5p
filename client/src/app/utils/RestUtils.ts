import axios from 'axios';

function buildUrl(path: string) {
  return `http://localhost:8080${path}`;
}

export async function postCall(
  path: string,
  data: any,
  headers: any
): Promise<any> {
  console.log('POST', buildUrl(path), JSON.stringify(data), headers);
  return await axios
    .post(buildUrl(path), JSON.stringify(data), {
      headers: headers || {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    .then((res) => {
      return res.data;
    })
    .catch((error) => {
      console.error(error);
    });
}
