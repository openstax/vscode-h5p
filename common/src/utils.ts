export function assertValue<T>(
  v: T | null | undefined,
  message = 'Expected a value but did not get anything',
) {
  if (v !== null && v !== undefined) return v;
  /* istanbul ignore next */
  throw new Error(`BUG: assertValue. Message: ${message}`);
}

export function assertType<T>(
  v: unknown,
  type: string,
  message = 'Got unexpected type',
): T {
  if (typeof v === type) return v as T;
  /* istanbul ignore next */
  throw new Error(`BUG: assertType. Message: ${message}`);
}

export function unwrap<T>(opt: T | null | undefined): T {
  return assertValue(opt, 'BUG: unwrap optional without value.');
}

export function chunk<T>(arr: T[], chunkSize: number): Array<T[]> {
  return [...range(Math.ceil(arr.length / chunkSize))]
    .map((n) => n * chunkSize)
    .map((offset) => arr.slice(offset, offset + chunkSize));
}

export function range(a: number, b?: number) {
  return b === undefined
    ? [...new Array(a)].map((_, idx) => idx)
    : [...new Array(b - a)].map((_, idx) => idx + a);
}

export function isFalsy<T>(obj: T): boolean {
  return (typeof obj === 'boolean' ? obj : Boolean(obj)) === false;
}

export function debounce<A = unknown, R = void>(
  fn: (args: A) => R,
  ms: number,
): (args: A) => Promise<R> {
  let timer: NodeJS.Timeout;

  return (args: A) => {
    clearTimeout(timer);

    return new Promise((resolve) => {
      timer = setTimeout(() => {
        resolve(fn(args));
      }, ms);
    });
  };
}

// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#converting_a_digest_to_a_hex_string
export async function digestMessage(
  message: string,
  options?: { algorithm?: string },
) {
  const { algorithm = 'SHA-1' } = options ?? {};
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await crypto.subtle.digest(algorithm, msgUint8); // hash the message
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray
    .map((b) => b.toString(16).padStart(2, '0'))
    .join(''); // convert bytes to hex string
  return hashHex;
}

export async function randomId(options?: {
  salt?: string;
  maxLength?: number;
  message?: string;
  algorithm?: string;
}) {
  const {
    salt = Date.now().toString(),
    maxLength = Infinity,
    message = '',
  } = options ?? {};
  const hash = await digestMessage(`${salt} ${message}`, options);
  return hash.slice(0, maxLength);
}
