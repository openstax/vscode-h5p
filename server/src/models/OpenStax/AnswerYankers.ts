export type Yanker = (content: any) => [unknown, unknown];
export type Unyanker = (publicData: any, privateData: any) => unknown;

function yankByKeys(content: any, keys: string[]): [unknown, unknown] {
  const publicData = {};
  const privateData = {};
  Object.entries(content).forEach(([k, v]) => {
    if (keys.includes(k)) {
      privateData[k] = v;
    } else {
      publicData[k] = v;
    }
  });
  return [publicData, privateData];
}

export const blanksYanker: Yanker = (content) => {
  return yankByKeys(content, ['questions']);
};

export const multiChoiceYanker: Yanker = (content) => {
  return yankByKeys(content, ['answers']);
};

export const questionSetYanker: Yanker = (content) => {
  const yankBySubtype = (q) => {
    const [libraryName, _version] = q.library.split(' ');
    switch (libraryName) {
      case 'H5P.MultiChoice':
        return multiChoiceYanker(q.params);
      case 'H5P.Blanks':
        return blanksYanker(q.params);
      case 'H5P.TrueFalse':
        return trueFalseYanker(q.params);
      default:
        throw new Error(`Library, "${libraryName}," is unsupported`);
    }
  };
  const privateData: unknown[] = [];
  const publicData = {
    ...content,
    // NOTE: Impure map
    questions: content.questions.map((q) => {
      const [pub, priv] = yankBySubtype(q);
      privateData.push(priv);
      return {
        ...q,
        params: pub,
      };
    }),
  };
  return [publicData, privateData];
};

export const questionSetMerge: Unyanker = (
  publicData: any,
  privateData: any
) => {
  const copy = { ...publicData };
  copy.questions.forEach((q, idx) => {
    q.params = shallowMerge(q.params, privateData[idx]);
  });
  return copy;
};

export const trueFalseYanker: Yanker = (content) => {
  return yankByKeys(content, ['correct']);
};

export const shallowMerge: Unyanker = (publicData: any, privateData: any) => ({
  ...publicData,
  ...privateData,
});
