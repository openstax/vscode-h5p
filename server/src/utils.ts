import * as os from 'os';
import decompress from 'decompress';
import fs from 'fs';
import express from 'express';
import { DOMParser } from '@xmldom/xmldom';
import * as xpath from 'xpath-ts';
import {
  assertTrue,
  assertValue,
  isFalsy,
  unwrap,
} from '../../common/src/utils';

export interface MergeOptions {
  customMerge?: (lhs: unknown, rhs: unknown, options?: MergeOptions) => unknown;
  arrayMerge?: (
    lhs: unknown[],
    rhs: unknown[],
    options?: MergeOptions,
  ) => unknown[];
}

/* istanbul ignore next (pure function that depends solely on express) */
export function createH5PRouter<EditRequestType, ContentRequestType>(
  onPlay: (req: any, res: any) => Promise<void>,
  onEdit: (req: EditRequestType, res: any) => Promise<void>,
  onNew: (req: ContentRequestType, res: any) => Promise<void>,
  onSave: (req: ContentRequestType, res: any) => Promise<void>,
  onDelete: (req: ContentRequestType, res: any) => Promise<void>,
  onFetch: (req: ContentRequestType, res: any) => Promise<void>,
): express.Router {
  const router = express.Router();

  router.get('/:contentId/play', onPlay);
  router.get('/:contentId/edit', onEdit);
  router.post('/', onNew);
  router.patch('/:contentId', onSave);
  router.delete('/:contentId', onDelete);
  router.get('/', onFetch);

  return router;
}

export function getIps(external: boolean = false): string[] {
  if (!external) return ['localhost'];
  const interfaces = os.networkInterfaces();
  return Object.values(interfaces).flatMap((devInts) =>
    assertValue(devInts)
      .filter((int: { internal: any }) => isFalsy(int.internal))
      .filter((int: { family: string }) => int.family === 'IPv4')
      .map((int: { address: any }) => int.address),
  );
}

export function fsRemove(folderPath: string) {
  fs.rmSync(folderPath, { recursive: true, force: true });
}

export async function extractArchive(
  path: string,
  destinationFolder: string,
  options?: {
    deleteArchive?: boolean;
    filesToExtract?: string[];
    decompressOptions?: decompress.DecompressOptions;
    verbose?: boolean;
  },
): Promise<void> {
  const {
    deleteArchive,
    filesToExtract,
    decompressOptions = { strip: 0 },
    verbose = true,
  } = options ?? {};
  console.log(`Extracting file ${path}`);
  try {
    const files = await decompress(path, destinationFolder, decompressOptions);
    if (filesToExtract !== undefined) {
      files
        .filter((file) => !filesToExtract.includes(file.path))
        .forEach((file) => fsRemove(file.path));
    }
    if (verbose) {
      (filesToExtract ?? files.map((file) => file.path)).forEach((name) => {
        console.log(`${destinationFolder}/${name}`);
      });
    }
    if (deleteArchive === true) {
      fsRemove(path);
    }
  } catch (e) {
    /* istanbul ignore next */
    console.error(e);
  }
}

export class ParseError extends Error {}

export function parseToDOM(xmlString: string, mimeType = 'application/xml') {
  const locator = { lineNumber: 0, columnNumber: 0 };
  /* istanbul ignore next */
  const cb = () => {
    const pos = {
      line: locator.lineNumber - 1,
      character: locator.columnNumber - 1,
    };
    throw new ParseError(`ParseError: ${JSON.stringify(pos)}`);
  };
  const p = new DOMParser({
    locator,
    errorHandler: {
      warning: console.warn,
      error: cb,
      fatalError: cb,
    },
  });
  const doc = p.parseFromString(xmlString, mimeType);
  return doc;
}

export function parseBooksXML(booksXmlPath: string): {
  booksRoot: string;
  pagesRoot: string;
  mediaRoot: string;
  privateRoot: string;
  publicRoot: string;
} {
  const doc = parseToDOM(fs.readFileSync(booksXmlPath, 'utf-8'));
  const select = xpath.useNamespaces({
    bk: 'https://openstax.org/namespaces/book-container',
  });
  const bookVars: Record<string, string> = Object.fromEntries(
    (select('//bk:var', doc) as Element[]).map((n) =>
      [n.getAttribute('name'), n.getAttribute('value')].map(unwrap).map((i) => {
        const trimmed = i.trim();
        if (trimmed.length === 0) {
          /* istanbul ignore next */
          throw new Error('Found empty var in books.xml');
        }
        return trimmed;
      }),
    ),
  );
  return {
    booksRoot: bookVars['BOOKS_ROOT'] ?? '/collections',
    pagesRoot: bookVars['PAGES_ROOT'] ?? '/modules',
    mediaRoot: bookVars['MEDIA_ROOT'] ?? '/media',
    privateRoot: bookVars['PRIVATE_ROOT'] ?? '/private',
    publicRoot: bookVars['PUBLIC_ROOT'] ?? '/interactives',
  };
}

export function mergeByIndex(
  lhs: unknown[],
  rhs: unknown[],
  options?: MergeOptions,
) {
  return Object.values(
    recursiveMerge(
      Object.fromEntries(Object.entries(lhs)),
      Object.fromEntries(Object.entries(rhs)),
      options,
    ) as object,
  );
}

export function defaultValueMerge(lhs: unknown, rhs: unknown) {
  return rhs ?? lhs;
}

export function recursiveMerge(
  lhs: unknown,
  rhs: unknown,
  options?: MergeOptions,
): unknown {
  const { arrayMerge = mergeByIndex, customMerge = defaultValueMerge } =
    options ?? {};
  if (
    typeof lhs === 'object' &&
    typeof rhs === 'object' &&
    lhs != null &&
    rhs != null
  ) {
    const isArrayL = Array.isArray(lhs);
    const isArrayR = Array.isArray(rhs);
    if (isArrayL || isArrayR) {
      assertTrue(isArrayL && isArrayR, 'Expected two arrays');
      return arrayMerge(lhs as unknown[], rhs as unknown[], options);
    } else {
      const keysL = Object.keys(lhs);
      const keysR = Object.keys(rhs);
      const sharedKeys = keysL.filter((k) => keysR.indexOf(k) !== -1);
      const merged = Object.fromEntries(
        sharedKeys.map((k) => [
          k,
          recursiveMerge(Reflect.get(lhs, k), Reflect.get(rhs, k), options),
        ]),
      );
      return { ...lhs, ...rhs, ...merged };
    }
  } else {
    return customMerge(lhs, rhs, options);
  }
}
