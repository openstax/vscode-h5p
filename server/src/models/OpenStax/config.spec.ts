import { recursiveMerge } from '../../utils';
import { yankByKeysFactory } from './AnswerYankers';
import Config, { newSupportedLibrary } from './config';

describe('questionSetYanker', () => {
  it('yanks private data by subtype', () => {
    const fakeContent = {
      questions: [
        {
          params: {
            question: '<p>Given point a and point b, what is 1+1?</p>\n',
            correct: false,
          },
          library: 'H5P.TrueFalse 1.16',
        },
        {
          params: {
            question: '<p>Given point a and point b, what is 2+2?</p>\n',
            answers: [
              {
                correct: false,
                tipsAndFeedback: {
                  tip: '',
                  chosenFeedback: '',
                  notChosenFeedback: '',
                },
                text: '<div>3</div>\n',
              },
              {
                correct: true,
                tipsAndFeedback: {
                  tip: '',
                  chosenFeedback: '',
                  notChosenFeedback: '',
                },
                text: '<div>4</div>\n',
              },
            ],
          },
          library: 'H5P.MultiChoice 1.16',
        },
        {
          params: {
            media: {
              disableImageZooming: false,
            },
            questions: ['<p>What is *true*?</p>\n'],
          },
          library: 'H5P.Blanks 1.14',
        },
      ],
    };
    const copy = JSON.parse(JSON.stringify(fakeContent));
    const questionSetYanker =
      Config.supportedLibraries['H5P.QuestionSet']?.yankAnswers;
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const [publicData, privateData] = questionSetYanker!(fakeContent);
    // No modifications to original
    expect(copy).toStrictEqual(fakeContent);
    expect(publicData).toMatchSnapshot();
    expect(privateData).toMatchSnapshot();
    expect(recursiveMerge(publicData, privateData)).toStrictEqual(fakeContent);
  });
});

describe('newSupportedLibrary', () => {
  it('combines library-specific and general answer yankers', () => {
    const fakeContent = {
      a: 1,
      b: 2,
      c: 3,
      d: 4,
    };
    // Test combination of additionalFields, privateKeys, and yankAnswers
    const testLib = newSupportedLibrary({
      yankAnswers: yankByKeysFactory('b', 'c'),
      semantics: {
        additionalFields: [
          { field: { name: 'a', type: 'text' }, private: true },
        ],
      },
    });
    const [publicData, privateData] = testLib.yankAnswers(fakeContent);
    expect(publicData).toStrictEqual({ d: 4 });
    expect(privateData).toStrictEqual({ a: 1, b: 2, c: 3 });
  });
  it('Uses whichever answer yanker is defined', () => {
    const fakeContent = {
      a: 1,
      b: 2,
      c: 3,
      d: 4,
    };
    // Test combination of additionalFields, privateKeys, and yankAnswers
    const testLib = newSupportedLibrary({
      semantics: {
        additionalFields: [
          { field: { name: 'a', type: 'text' }, private: true },
        ],
      },
    });
    let [publicData, privateData] = testLib.yankAnswers(fakeContent);
    expect(publicData).toStrictEqual({ b: 2, c: 3, d: 4 });
    expect(privateData).toStrictEqual({ a: 1 });
    const testLib2 = newSupportedLibrary({
      yankAnswers: yankByKeysFactory('a'),
    });
    [publicData, privateData] = testLib2.yankAnswers(fakeContent);
    expect(publicData).toStrictEqual({ b: 2, c: 3, d: 4 });
    expect(privateData).toStrictEqual({ a: 1 });
  });
});
