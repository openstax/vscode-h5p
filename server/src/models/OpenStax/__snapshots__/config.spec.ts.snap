// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`questionSetYanker yanks private data by subtype 1`] = `
{
  "questions": [
    {
      "library": "H5P.TrueFalse 1.16",
      "params": {
        "question": "<p>Given point a and point b, what is 1+1?</p>
",
      },
    },
    {
      "library": "H5P.MultiChoice 1.16",
      "params": {
        "question": "<p>Given point a and point b, what is 2+2?</p>
",
      },
    },
    {
      "library": "H5P.Blanks 1.14",
      "params": {
        "media": {
          "disableImageZooming": false,
        },
      },
    },
  ],
}
`;

exports[`questionSetYanker yanks private data by subtype 2`] = `
{
  "questions": [
    {
      "params": {
        "correct": false,
      },
    },
    {
      "params": {
        "answers": [
          {
            "correct": false,
            "text": "<div>3</div>
",
            "tipsAndFeedback": {
              "chosenFeedback": "",
              "notChosenFeedback": "",
              "tip": "",
            },
          },
          {
            "correct": true,
            "text": "<div>4</div>
",
            "tipsAndFeedback": {
              "chosenFeedback": "",
              "notChosenFeedback": "",
              "tip": "",
            },
          },
        ],
      },
    },
    {
      "params": {
        "questions": [
          "<p>What is *true*?</p>
",
        ],
      },
    },
  ],
}
`;
